isomorphe/geonetwork.py

```diff
@@ -343,6 +364,86 @@ def get_groups(self) -> dict:
         return r.json()
 
 
+class GeonetworkClientV3(GeonetworkClient):
+    version = 3
+
+    def _search_params(self, query: dict[str, Any] | None) -> dict[str, Any]:
+        params = {
+            "_content_type": "json",
+            "buildSummary": "false",
+            "fast": "index",  # needed to get info such as title
+            "sortBy": "changeDate",
+            "sortOrder": "reverse",
+        }
+        if query:
+            params |= query
+        return params
+
+    def _search_hits(self, params: dict[str, Any], from_pos: int) -> list[dict]:
+        r = self.session.get(
+            f"{self.api}/q",
+            headers={"Accept": "application/json"},
+            params=params | {"from": from_pos},
+        )
+        r.raise_for_status()
+        rsp = r.json()
+        hits = rsp.get("metadata")
+        if hits and "geonet:info" in hits:
+            # When returning a single record, metadata isn't a list :/
+            hits = [hits]
+        return hits
+
+    def _as_record(self, hit: dict) -> Record:
+        return Record(
+            uuid=hit["geonet:info"]["uuid"],
+            title=hit.get("defaultTitle"),
+            md_type=self._get_metadata_type(hit),
+            state=self._get_workflow_state(hit),
+        )
+
+
+class GeonetworkClientV4(GeonetworkClient):
+    version = 4
+
+    def _search_params(self, query: dict[str, Any] | None) -> dict[str, Any]:
+        data = {
+            "size": 20,
+            "sort": [{"changeDate": "asc"}],
+            "_source": [
+                "uuid",
+                "resourceTitleObject.default",
+                "resourceType",
+                "draft",
+                "isTemplate",
+                "mdStatus",
+            ],
+        }
+        if query:
+            q = " ".join(f"+{k}:{v}" for k, v in query.items())
+            data |= {"query": {"bool": {"filter": [{"query_string": {"query": q}}]}}}
+        return data
+
+    def _search_hits(self, params: dict[str, Any], from_pos: int) -> list[dict]:
+        r = self.session.post(
+            f"{self.api}/search/records/_search",
+            headers={"Accept": "application/json", "Content-Type": "application/json"},
+            params={"bucket": "metadata"},
+            json=params | {"from": from_pos},
+        )
+        r.raise_for_status()
+        hits = r.json().get("hits") or {}
+        return hits.get("hits")  # nested "hits"
+
+    def _as_record(self, hit: dict) -> Record:
+        md = hit.get("_source")
+        return Record(
+            uuid=md.get("uuid"),
+            title=md.get("resourceTitleObject", {}).get("default"),
+            md_type=self._get_metadata_type(md),
+            state=self._get_workflow_state(md),
+        )
```

Comment:
```suggestion
    def _as_record(self, hit: dict) -> Record:
        md = hit.get("_source") or {}
        return Record(
            uuid=md.get("uuid") or "",
            title=md.get("resourceTitleObject", {}).get("default"),
            md_type=self._get_metadata_type(md),
            state=self._get_workflow_state(md),
        )
```

Maybe use a better type for hit.

Reply:
Best I can do is `dict[str,Any]`. I'll add it but it's pretty useless...